// @remotion/studio/dist/components/WebRender/TriggerWebRender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerWebRender = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = require("react");
const remotion_1 = require("remotion");
const in_out_1 = require("../../state/in-out");
const modals_1 = require("../../state/modals");
const Button_1 = require("../Button");
const button = {
    paddingLeft: 7,
    paddingRight: 7,
    paddingTop: 7,
    paddingBottom: 7,
};
const label = {
    fontSize: 14,
};
const TriggerWebRender = () => {
    const video = remotion_1.Internals.useVideo();
    const getCurrentFrame = player_1.PlayerInternals.useFrameImperative();
    const { inFrame, outFrame } = (0, in_out_1.useTimelineInOutFramePosition)();
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const onClick = (0, react_1.useCallback)(() => {
        if (!(video === null || video === void 0 ? void 0 : video.id)) {
            return null;
        }
        const frame = getCurrentFrame();
        setSelectedModal({
            type: 'web-render',
            initialFrame: frame,
            compositionId: video.id,
            defaultProps: video.defaultProps,
            inFrameMark: inFrame,
            outFrameMark: outFrame,
        });
    }, [
        getCurrentFrame,
        inFrame,
        outFrame,
        setSelectedModal,
        video === null || video === void 0 ? void 0 : video.defaultProps,
        video === null || video === void 0 ? void 0 : video.id,
    ]);
    if (!video) {
        return null;
    }
    return ((0, jsx_runtime_1.jsx)(Button_1.Button, { onClick: onClick, buttonContainerStyle: button, disabled: false, children: (0, jsx_runtime_1.jsx)("span", { style: label, children: "Render on web" }) }));
};
exports.TriggerWebRender = TriggerWebRender;
